src/include/miopen/anyramdb.hpp

```diff
@@ -0,0 +1,141 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/db.hpp>
+#include <miopen/db_record.hpp>
+#include <miopen/lock_file.hpp>
+
+#include <boost/optional.hpp>
+#include <boost/any.hpp>
+
+#include <chrono>
+#include <map>
+#include <string>
+#include <sstream>
+
+namespace miopen {
+
+class LockFile;
+
+struct AnyRamDb
```

Comment:
Is there any *noticeable* performance difference between this and RamDb? Especially in miopen API call time rather than just in db measure printed nanoseconds. If so, is there any reason for not updating RamDb rather than duplicating it?

Reply:
_On behalf of @JehandadKhan:_
> @atamazov You mentioned earlier that this file is very similar to ramdb etc, which is correct. However, note that this file presents a version of the database which hold an object instead of a string type which is why this class was realized. If you have a better suggestion please share it.

Reply:
@JehandadKhan What I am thinking is that instead of copying-and-modification we should always try to generalize at first. In this specific case, we can try to generalize RamDB to, er, AnyRamDb (or GenericRamDb). The idea is to make it parametrized with payload type, so we can instantiate it with string (and get old RamDb behavior) or whatever else (and get ram db class that is necessary for AI heuristic).

Reply:
I mean there is no need to generalize at all if there is no difference in performance between this and RambDb.

Reply:
@atamazov I tried templating the classes that AnyDamDb implements (FindRecord, StoreRecord, RemoveRecord) within RamDb to find that it was not a trivial piece of work to implement the functionality of AnyRamDb while maintaining the original behavior of RamDb. This was mainly due to the fact that RamDB (and DbRecord) expects the objects that it stores to have the serialize and deserialize (parsing) functions defined for them. I believe there would need to be templated versions of some functions (for objects that do not require serialization) and then a non-templated hard typed function (overload the same function basically) that would maintain the original behavior of RamDb. Also the objects that AnyRamDb stores do not need to be stored in a text database, it only maintains it's cache throughout a single MIOpen run. I have created a ticket so that either Evgenii/Vasilii/or myself can look at this in the future, for now can you accept this implementation?

Reply:
@Dmantri98 I would prefer to avoid postponing this problem. I do believe you can do the necessary code refactoring, may with some help from @DrizztDoUrden. 

Vasilii, can you please suggest how to properly generalize ram db?

Reply:
@atamazov  I believe we can deal with the design questions in another PR, smaller but frequent changes, while each point in develop should be valid :)

Reply:
@junliume If @Dmantri98 agrees to fix this issue in his literally next PR, then I won't mind.

Reply:
We can decide who does it later, however, we have created an issue to track the work.

Reply:
[off-topic] @JehandadKhan @junliume I believe the rule of thumb is that the engineer who "creates" the problem is supposed to be responsible for resolving it.

A curious consequence: an engineer that "initiates" a problem that he/she is currently unable to resolve is implicitly inspired to improve own skills/knowledge.